url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/Node.java

```diff
@@ -0,0 +1,1030 @@
+		if (index < 0 || index >= m_children)
+			throw new IndexOutOfBoundsException("" + index);
+
+		return m_pIdentifier[index];
+	}
+
+	@Override
+	public IShape getChildShape(int index) throws IndexOutOfBoundsException {
+		if (index < 0 || index >= m_children)
+			throw new IndexOutOfBoundsException("" + index);
+
+		return new Region(m_pMBR[index]);
+	}
+
+	@Override
+	public int getLevel() {
+		return m_level;
+	}
+
+	@Override
+	public boolean isLeaf() {
+		return (m_level == 0);
+	}
+
+	@Override
+	public boolean isIndex() {
+		return (m_level != 0);
+	}
+
+	//
+	// Internal
+	//
+
+	protected Node(RTree pTree, int id, int level, int capacity) {
+		m_pTree = pTree;
+		m_level = level;
+		m_identifier = id;
+		m_capacity = capacity;
+		m_nodeMBR = (Region) pTree.m_infiniteRegion.clone();
+
+		m_pDataLength = new int[m_capacity + 1];
+		m_pData = new byte[m_capacity + 1][];
+		m_pMBR = new Region[m_capacity + 1];
+		m_pIdentifier = new int[m_capacity + 1];
+	}
+
+	/**
+	 * insert a entry into this node, child_num++
+	 * 
+	 * @param pData
+	 * @param mbr
+	 * @param id
+	 * @throws IllegalStateException
+	 */
+	protected void insertEntry(byte[] pData, Region mbr, int id) throws IllegalStateException {
+		if (m_children >= m_capacity)
+			throw new IllegalStateException("m_children >= m_nodeCapacity");
+
+		m_pDataLength[m_children] = (pData != null) ? pData.length : 0;
+		// TODO Maybe this step should change to copy instead of point to.
+		m_pData[m_children] = pData;
+		m_pMBR[m_children] = mbr;
+		m_pIdentifier[m_children] = id;
+
+		m_totalDataLength += m_pDataLength[m_children];
+		m_children++;
+
+		Region.combinedRegion(m_nodeMBR, mbr);
+	}
+
+	protected void deleteEntry(int index) throws IndexOutOfBoundsException {
+		if (index < 0 || index >= m_children)
+			throw new IndexOutOfBoundsException("" + index);
+
+		boolean touches = m_nodeMBR.touches(m_pMBR[index]);
+
+		m_totalDataLength -= m_pDataLength[index];
+		m_pData[index] = null;
+
+		if (m_children > 1 && index != m_children - 1) {
+			//replace the target entry with last enrty, then m_children--
+			m_pDataLength[index] = m_pDataLength[m_children - 1];
+			m_pData[index] = m_pData[m_children - 1];
+			m_pData[m_children - 1] = null;
+			m_pMBR[index] = m_pMBR[m_children - 1];
+			m_pMBR[m_children - 1] = null;
+			m_pIdentifier[index] = m_pIdentifier[m_children - 1];
+		}
+
+		m_children--;
+
+		if (m_children == 0) {
+			// means that no child remains
+			m_nodeMBR = (Region) m_pTree.m_infiniteRegion.clone();
+		} else if (touches) {
+			for (int cDim = 0; cDim < m_pTree.m_dimension; cDim++) {
+				m_nodeMBR.m_pLow[cDim] = Double.POSITIVE_INFINITY;
+				m_nodeMBR.m_pHigh[cDim] = Double.NEGATIVE_INFINITY;
+
+				for (int cChild = 0; cChild < m_children; cChild++) {
+					m_nodeMBR.m_pLow[cDim] = Math.min(m_nodeMBR.m_pLow[cDim], m_pMBR[cChild].m_pLow[cDim]);
+					m_nodeMBR.m_pHigh[cDim] = Math.max(m_nodeMBR.m_pHigh[cDim], m_pMBR[cChild].m_pHigh[cDim]);
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 238:246, 322:330.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.